prov/gni/src/gnix_vc.c

```diff
@@ -69,6 +70,367 @@ static int __gnix_vc_conn_ack_comp_fn(void *data);
+			} else {
+				ret = _gnix_map_lookup(av,
+						       *((fi_addr_t *)dest_addr),
+						       av_entry);
+                                assert(*av_entry);
+                                GNIX_DEBUG(FI_LOG_EP_CTRL,
+					   "av_entry at %p in __gnix_vc_lookup_vc",
+					   *av_entry);
+
+                                if (ret != FI_SUCCESS) {
+					GNIX_WARN(FI_LOG_EP_CTRL,
+						  "_gnix_map_lookup returned %s\n",
+						  fi_strerror(-ret));
+					return ret;
+				}
+				memcpy(ht_dest_addr, &((av_entry[0])->gnix_addr),
+				       sizeof(struct gnix_address));
+
+				key_ptr = malloc(sizeof(gnix_ht_key_t));
+				assert(key_ptr);
+
+                                memcpy(key_ptr, ht_dest_addr,
+                                       sizeof(gnix_ht_key_t));
+
+                                GNIX_INFO(FI_LOG_EP_CTRL,
+					  "fi_addr_t: 0x%llx gnix_addr: 0x%llx\n",
+					  *((fi_addr_t *) dest_addr),
+					  (av_entry[0])->gnix_addr);
+			}
+
+		}
+		/* fastlock_acquire(&ep->vc_lock); */
+		*vc_ptr = (struct gnix_vc *)_gnix_ht_lookup(ep->vc_ht, *key_ptr);
+		/* fastlock_release(&ep->vc_lock); */
+		*key = key_ptr;
+                GNIX_DEBUG(FI_LOG_EP_CTRL,
+                           "Found vc %p at hashtable key %llu\n", *vc_ptr,
+                           *key_ptr);
+                return ret;
+	} else {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Invalid FI_AV type of \"%d\" in __gnix_vc_lookup_vc",
+			  av_type);
+		return -FI_EINVAL;
+	}
+}
+
+/**
+ * Look up the vc, if it's found just return it, otherwise allocate a new vc,
+ * insert it into the hashtable or vector, and connect it.
+ *
+ * @assumption: ep is non-null;
+ * dest_addr is valid;
+ * is_dest_addr_gnix_addr is correct for the given dest_addr;
+ * vc_ptr is non-null.
+ */
+static inline int __gnix_vc_get_vc(struct gnix_fid_ep *ep, void *dest_addr,
+				   bool is_dest_addr_gnix_addr,
+				   struct gnix_vc **vc_ptr)
+{
+	struct gnix_fid_av *av = ep->av;
+	assert(av);
+	int av_type = av->type, ret = FI_SUCCESS;
+	struct gnix_av_addr_entry *av_entry = NULL;
+	gnix_ht_key_t *key_ptr = NULL;
+	struct gnix_vc *vc_tmp = NULL;
+	fi_addr_t table_dest_addr;
+	struct gnix_address ht_dest_addr;
+
+	GNIX_TRACE(FI_LOG_EP_CTRL, "\n");
+        GNIX_DEBUG(FI_LOG_EP_CTRL,
+                   "ep->vc_table = %p, ep->vc_table->vector = %p\n",
+                   ep->vc_table, ep->vc_table->vector);
+
+        fastlock_acquire(&ep->vc_lock);
+
+	/*
+         * if AV type is FI_AV_TABLE, table_dest_addr will be filled in;
+         * otherwise
+         * if AV type is FI_AV_MAP, ht_dest_addr and key will be filled in.
+         */
+        ret = __gnix_vc_lookup_vc(ep, dest_addr, is_dest_addr_gnix_addr,
+                                  &av_entry, &key_ptr, &table_dest_addr,
+                                  &ht_dest_addr, &vc_tmp);
+
+        /* Ignore FI_ECANCELED as it's returned when looking up an empty entry */
+	if (ret != -FI_ECANCELED && ret != FI_SUCCESS) {
```

Comment:
When _gnix_vec_at finds that the passed index refers to an empty (NULL) entry.
> On Mar 28, 2016, at 9:00 AM, Howard Pritchard <notifications@github.com> wrote:
> 
> In prov/gni/src/gnix_vc.c <https://github.com/ofi-cray/libfabric-cray/pull/736#discussion_r57580650>:
> 
> > +                   "ep->vc_table = %p, ep->vc_table->vector = %p\n",
> > +                   ep->vc_table, ep->vc_table->vector);
> > +
> > +        fastlock_acquire(&ep->vc_lock);
> > +
> > +	/*
> > +         * if AV type is FI_AV_TABLE, table_dest_addr will be filled in;
> > +         * otherwise
> > +         * if AV type is FI_AV_MAP, ht_dest_addr and key will be filled in.
> > +         */
> > +        ret = __gnix_vc_lookup_vc(ep, dest_addr, is_dest_addr_gnix_addr,
> > +                                  &av_entry, &key_ptr, &table_dest_addr,
> > +                                  &ht_dest_addr, &vc_tmp);
> > +
> > +        /* Ignore FI_ECANCELED as it's returned when looking up an empty entry */
> > +	if (ret != -FI_ECANCELED && ret != FI_SUCCESS) {
> when will -FI_ECANCELD be returned?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub <https://github.com/ofi-cray/libfabric-cray/pull/736/files/408e2a2ddff4bc9751170dc9ac6f6a66a970679b#r57580650>